### PR TITLE
feat: add types

### PR DESCRIPTION
**Pull request by achingbrain**
Adds ts type generation and fixes all tsc errors.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,14 +29,15 @@
...
+    "prepare": "aegir build --no-bundle",
     "test:node": "aegir test --target node",
     "test:browser": "aegir test --target browser",
     "test": "aegir test",
     "lint": "aegir lint && aegir lint-package-json",
     "release": "aegir release",
     "release-minor": "aegir release --type minor",
     "release-major": "aegir release --type major",
-    "build": "aegir build",
+    "prepublishOnly": "aegir build",
```
Comment (hugomrdias) [inline package.json:40]: why do we need this ?

File: package.json [hunk 2]
```diff
@@ -29,14 +29,15 @@
...
     "coverage": "aegir coverage",
     "coverage-publish": "aegir coverage --upload"
```
Comment (hugomrdias) [inline package.json:42]: we should remove these, aegir doesnt have them for a long time

**Review by hugomrdias — CHANGES_REQUESTED** [review]

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "./node_modules/aegir/src/config/tsconfig.aegir.json",
```
Comment (Gozala) [inline tsconfig.json:2]: Let's not assume where exactly aegir is.

```suggestion
  "extends": "aegir/src/config/tsconfig.aegir.json",
```

**Review by Gozala — APPROVED** [review]
Just a minor suggestion to avoid hard coding aegir path. Otherwise looks good to me.

File: tsconfig.json (hunk 3 above)
Comment (hugomrdias) [reply tsconfig.json:2]: That started failing recently and we don't know why yet.

**Comment by Gozala** [comment]
It might be worth considering changing some of the `isFoo` style functions to more type friendly ones like `asFoo(x:any): Foo|null` so that type checker is able to refine the type. Unfortunately isFoo(x:any): x is Foo` is not applicable here because I do not believe these functions check if passed value is `Foo` but test whether value could potentially be casted to `Foo`.

File: tsconfig.json (hunk 3 above)
Comment (achingbrain) [reply tsconfig.json:2]: The typescript team broke this a minor release last week:

```
> is-ipfs@3.0.0 prepare
> aegir build --no-bundle

[11:11:07] Clean ./dist [started]
[11:11:07] Clean ./dist [completed]
[11:11:07] Generate types [started]
tsconfig-types.aegir.json:1:12 - error TS6053: File 'aegir/src/config/tsconfig.aegir.json' not found.
```

File: package.json (hunk 1 above)
Comment (achingbrain) [reply package.json]: To publish a minified bundle?  Have removed it, shout if it's actually useful.

File: package.json (hunk 2 above)
Comment (achingbrain) [reply package.json]: Done.

File: src/index.js [hunk 4]
```diff
@@ -5,6 +5,7 @@ const multibase = require('multibase')
 const Multiaddr = require('multiaddr')
 const mafmt = require('mafmt')
 const CID = require('cids')
+// @ts-ignore
```
Comment (hugomrdias) [inline src/index.js:8]: i released iso-url v1.1.0 with types

**Review by hugomrdias — CHANGES_REQUESTED** [review]

**Review by hugomrdias — APPROVED** [review]

File: .travis.yml [hunk 5]
```diff
@@ -1,10 +1,16 @@
...
 stages:
   - check
   - test
   - cov
 
+branches:
+  only:
+  - master
+  - /^release\/.*$/
+
```
Comment (lidel) [inline .travis.yml:13]: Won't this  disable tests on PRs?

File: package.json [hunk 6]
```diff
@@ -29,28 +30,26 @@
     "url": "https://github.com/ipfs/is-ipfs.git"
   },
   "scripts": {
+    "prepare": "aegir build --no-bundle",
```
Comment (lidel) [inline package.json:33]: What does `--no-bundle` do?

**Review by lidel — COMMENTED** [review]
@Gozala  I'd argue that the purpose of this library is to validate strings/bytes and return clear booleans specifically to avoid false-negatives when comparing types in (non-TS) userland. At least that is why I started using it in ipfs-companion (wanted deterministic `CID.isCID` i could trust)

Anyway, adding types sounds sensible, are we good with this as it is now, or should we wait till 
```
tsconfig-types.aegir.json:1:12 - error TS6053: File 'aegir/src/config/tsconfig.aegir.json' not found.
```
Gets resolved upstream?

File: .travis.yml (hunk 5 above)
Comment (achingbrain) [reply .travis.yml:13]: No - see the checks that have run on this PR.

If you don't specify this they run twice - once for the PR and once for the branch.

File: package.json (hunk 6 above)
Comment (achingbrain) [reply package.json:33]: It skips the browser bundle step as you don't need it during development, i.e. it only generates types from the jsdoc comments.

**Comment by achingbrain** [comment]
> tsconfig-types.aegir.json:1:12 - error TS6053: File 'aegir/src/config/tsconfig.aegir.json' not found.

This would only start happening if we apply the suggestion from https://github.com/ipfs-shipyard/is-ipfs/pull/39#discussion_r583939185 which we should not do.

I would not wait on it getting resolved first, the ts team move at their own cadence with their own conventions.

File: .travis.yml (hunk 5 above)
Comment (lidel) [reply .travis.yml:13]: Ah, Thanks! I  forgot this is Travis, not GoogleActions :man_facepalming:

File: package.json (hunk 1 above)
Comment (lidel) [reply package.json]: Did I miss any policy change for our JS libs? Why would we NOT provide a prebuilt browser bundle?

It is documented in https://github.com/ipfs-shipyard/is-ipfs#in-the-browser-through-script-tag  so for now I've restored `prepublishOnly`  in 2a03f54

**Review by lidel — APPROVED** [review]
lgtm 

(I've restored browser bundle in 2a03f54 because its still in [readme](https://github.com/ipfs-shipyard/is-ipfs#in-the-browser-through-script-tag))

**Comment by lidel** [comment]
This shipped as [is-ipfs@4.0.0](https://github.com/ipfs-shipyard/is-ipfs/releases/tag/v4.0.0)

**Pull request merged**